hal-x86_64/src/cpu/mod.rs

```diff
@@ -26,22 +26,22 @@ impl Port {
     /// Reading from a CPU port is unsafe.
     pub unsafe fn readb(&self) -> u8 {
         let result: u8;
-        llvm_asm!("in al, dx" : "={al}"(result) : "{dx}"(self.num) :: "volatile", "intel");
+        asm!("in al, dx", in("dx") self.num, out("al") result);
```

Comment:
also, i _think_ all the read/write intrinsics can have `options(readonly)`?

Reply:
> i suspect we could probably change these to
> and similar, rather than using `al` and `dx` explicitly...that _might_ be friendlier to the optimizer but also probably doesn't matter?

I'm not sure if we can. The `in` instruction is not documented as accepting arbitrary registers, but rather as accepting different forms of the A register as the first argument, and an immediate or different versions of the D register as the second argument: https://c9x.me/x86/html/file_module_x86_id_139.html

> also, i _think_ all the read/write intrinsics can have `options(readonly)`?

I wasn't sure whether or not that would be correct, but probably? I tended towards not adding extra `options(...)` constraints unless I was very confident that it would be correct to do so.

In this case, the previous code had marked the `asm` block as "volatile", which made me think that it probably couldn't be marked as `readonly`.

Reply:
i dunno why the `asm` block was marked volatile, though i see in the state before it was `llvm_asm`'d it was also `volatile`? `in` and `out` don't touch memory so these can be `readonly`.

Reply:
i have no idea why this was marked volatile? _maybe_ there was a concern about the read being optimized out if the result is never consumed, or something?